### PR TITLE
lighthouse: init at 2.5.1

### PR DESCRIPTION
**Pull request by philipmw**
###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/blockchains/lighthouse/default.nix [hunk 1]
```diff
@@ -0,0 +1,75 @@
+{ fetchFromGitHub, fetchurl, lib, rustPlatform, stdenv,
+  clang, cmake, darwin, llvmPackages, perl, protobuf }:
```
Comment (ocfox) [inline pkgs/applications/blockchains/lighthouse/default.nix:2]:
```suggestion
{ lib,
  perl,
  fetchurl,
  stdenv,
  clang,
  cmake,
  darwin,
  protobuf,
  rustPlatform,
  llvmPackages,
  fetchFromGitHub,
}:
```

File: maintainers/maintainer-list.nix [hunk 2]
```diff
@@ -10298,6 +10298,14 @@
     githubId = 178496;
     name = "Philipp Middendorf";
   };
+  pmw = {
+    email = "philip@mailworks.org";
+    matrix = "@philip4g:matrix.org";
+    name = "Philip White";
```
Comment (ocfox) [inline maintainers/maintainer-list.nix:10304]:
```suggestion
    name = "Philip White";
    github = "philipmw";
    githubId = 1379645;
```

File: pkgs/top-level/all-packages.nix [hunk 3]
```diff
@@ -36339,7 +36339,5 @@ with pkgs;
 
   mictray = callPackage ../tools/audio/mictray { };
 
-  swift-corelibs-libdispatch = callPackage ../development/libraries/swift-corelibs-libdispatch { };
-
-  swaysettings = callPackage ../applications/misc/swaysettings { };
```
Comment (ocfox) [inline pkgs/top-level/all-packages.nix:36344]: Why did you remove other packages?

File: pkgs/applications/blockchains/lighthouse/default.nix [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+  };
+
+  LIGHTHOUSE_DEPOSIT_CONTRACT_SPEC_URL = "file://${safe_contracts_file}";
+  LIGHTHOUSE_DEPOSIT_CONTRACT_TESTNET_URL = "file://${unsafe_contracts_file}";
+  meta = with lib; {
+    description = "An open-source Ethereum consensus client, written in Rust and maintained by Sigma Prime.";
+    homepage = "https://github.com/sigp/lighthouse";
+    license = licenses.asl20;
+    maintainers = with maintainers; [ pmw ];
+  };
```
Comment (ocfox) [inline pkgs/applications/blockchains/lighthouse/default.nix:66]: Please move this part to the end.

File: pkgs/applications/blockchains/lighthouse/default.nix [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+  clang, cmake, darwin, llvmPackages, perl, protobuf }:
+
+rustPlatform.buildRustPackage rec {
+  pname = "lighthouse";
+  version = "v2.5.1";
+
+  src = fetchFromGitHub {
+    owner = "sigp";
+    repo = pname;
+    rev = version;
```
Comment (ocfox) [inline pkgs/applications/blockchains/lighthouse/default.nix:11]:
```suggestion
  version = "2.5.1";

  src = fetchFromGitHub {
    owner = "sigp";
    repo = pname;
    rev = "v${version}";
```

**Review by ocfox — CHANGES_REQUESTED** [review]

File: pkgs/applications/blockchains/lighthouse/default.nix (hunk 4 above)
Comment (ocfox) [reply pkgs/applications/blockchains/lighthouse/default.nix:66]: Also, it would be nice if the description could be shortened, and it does not need to end with a period.

File: pkgs/top-level/all-packages.nix (hunk 3 above)
Comment (philipmw) [reply pkgs/top-level/all-packages.nix]: Mistake. Fixed.

File: maintainers/maintainer-list.nix [hunk 6]
```diff
@@ -10298,6 +10298,16 @@
     githubId = 178496;
     name = "Philipp Middendorf";
   };
+  pmw = {
```
Comment (ocfox) [inline maintainers/maintainer-list.nix:10301]: Please extract this change to a separate commit, see #182036 for an example.

File: pkgs/top-level/all-packages.nix [hunk 7]
```diff
@@ -36342,4 +36342,6 @@ with pkgs;
   swift-corelibs-libdispatch = callPackage ../development/libraries/swift-corelibs-libdispatch { };
 
   swaysettings = callPackage ../applications/misc/swaysettings { };
+
+  lighthouse = callPackage ../applications/blockchains/lighthouse { };
```
Comment (ocfox) [inline pkgs/top-level/all-packages.nix:36346]: Please (roughly) alphabetically sort this, don't toss it at the bottom of the file.

**Review by ocfox — CHANGES_REQUESTED** [review]

File: pkgs/applications/blockchains/lighthouse/default.nix [hunk 8]
```diff
@@ -0,0 +1,86 @@
+{ lib,
```
Comment (viraptor) [inline pkgs/applications/blockchains/lighthouse/default.nix:1]: The usual formatting is:
```
{ lib
, perl
, fetchurl
...
```

Comment (ocfox) [reply pkgs/applications/blockchains/lighthouse/default.nix:1]: Sorry, I misremembered this. :/

**Comment by philipmw** [comment]
I rebased on latest master. Any more feedback?

**Review by ocfox — APPROVED** [review]
LGTM

**Comment by philipmw** [comment]
Can anyone speak to what's holding up this PR from getting merged?

File: pkgs/applications/blockchains/lighthouse/default.nix [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+  LIGHTHOUSE_DEPOSIT_CONTRACT_SPEC_URL = "file://${safe_contracts_file}";
+  LIGHTHOUSE_DEPOSIT_CONTRACT_TESTNET_URL = "file://${unsafe_contracts_file}";
+
+  # The following logic provides the Web3signer to avoid a network download in
+  # https://github.com/sigp/lighthouse/blob/stable/testing/web3signer_tests/build.rs
+  LIGHTHOUSE_WEB3SIGNER_VERSION = "22.8.0"; # `tag_name` from https://api.github.com/repos/ConsenSys/web3signer/releases/latest
+  web3signer_file = fetchurl {
+    url = "https://artifacts.consensys.net/public/web3signer/raw/names/web3signer.zip/versions/${LIGHTHOUSE_WEB3SIGNER_VERSION}/web3signer-${LIGHTHOUSE_WEB3SIGNER_VERSION}.zip";
+    sha256 = "sha256-4Fjx8iZEqAzym8EAT8L5TI7aT1I/1IhNURHiX1MADPE=";
+  };
```
Comment (lovesegfault) [inline pkgs/applications/blockchains/lighthouse/default.nix:78]: You don't seem to use this anywhere else here, how does the build find out that this file is located in the nix store?

File: pkgs/applications/blockchains/lighthouse/default.nix [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+    perl
+    protobuf
+  ];
+
+  # https://discourse.nixos.org/t/big-sur-frameworks-issue/10810
+  # https://github.com/nix-community/nixpkgs-fmt/issues/7
+  buildInputs = lib.optionals stdenv.isDarwin [
+    darwin.apple_sdk.frameworks.Security
+  ];
+
```
Comment (centromere) [inline pkgs/applications/blockchains/lighthouse/default.nix:50]: Should the following also be set?:
```
buildFeatures = [ "modern" "gnosis" ];
```

File: pkgs/applications/blockchains/lighthouse/default.nix [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+  cargoBuildFlags = [
+    # Make only the `lighthouse` executable. Without this, the package produces 7 executables.
+    "-p lighthouse"
+  ];
+
+  nativeBuildInputs = [
+    # https://lighthouse-book.sigmaprime.io/installation-source.html
+    clang
+    cmake
+    llvmPackages.libclang
```
Comment (centromere) [inline pkgs/applications/blockchains/lighthouse/default.nix:40]: I am not certain `llvmPackages.libclang` is required here, since it is referenced below.

File: pkgs/applications/blockchains/lighthouse/default.nix (hunk 9 above)
Comment (centromere) [reply pkgs/applications/blockchains/lighthouse/default.nix:78]: If the file is not present, `build.rs` will attempt to download it. If the download fails, the build breaks. I was able to overcome this by simply adding `"--exclude" "web3signer_tests"` to `cargoBuildFlags`.

**Review by centromere — CHANGES_REQUESTED** [review]

**Comment by centromere** [comment]
Also note that `3.0.0` has been [released](https://github.com/sigp/lighthouse/releases/tag/v3.0.0).
